src/transaction.rs

```diff
@@ -0,0 +1,220 @@
+use secp256k1::{PublicKey, Signature};
+use crate::{
+    slip::{Slip},
+    time::{create_timestamp}
+};
+
+/// A single record used in the history of transactions being routed around the network
+#[derive(Debug, PartialEq, Clone, Copy)]
+pub struct Hop {
+    pub from: PublicKey,
+    pub to: PublicKey,
+    sig: Signature,
+}
+
+impl Hop {
+    /// Creates a new `Hop`
+    ///
+    /// `to` - `secp256k1::PublicKey` address of where the transaction is headed
+    /// * `from` - `secp256k1::PublicKey` address of where the transaction came from
+    /// * `sign` - `secp256k1::Signature` verifying work done by routers
+    pub fn new(to: PublicKey, from: PublicKey, sig: Signature) -> Hop {
+        return Hop { to, from, sig }
+    }
+}
+/// A record containging data of funds between transfered between public addresses. It
+/// contains additional information as an optinal message field to transfer data around the network
+#[derive(Debug, PartialEq)]
+pub struct Transaction {
+    body: TransactionBody
+}
+
+/// Core data to be serialized/deserialized of `Transaction`
+#[derive(Debug, PartialEq)]
+pub struct TransactionBody {
+    id:       u64,
+    ts:       u64,
+    pub to:   Vec<Slip>,
+    pub from: Vec<Slip>,
+    sig:      Signature,
+    pub typ:  TransactionBroadcastType,
+    path:     Vec<Hop>,
+    pub msg:  Vec<u8>,
+}
+/// Enumerated types of `Transaction`s to be handlded by consensus
+#[derive(Debug, PartialEq, Clone)]
+pub enum TransactionBroadcastType {
```

Comment:
I'm confused why these types are included in both slip and tx. I think these are only needed at the tx level or they should be split into two. e.g. Tx are Normal, GoldenTicket, GoldenChunk Rebroadcast and Slips are Normal or Fee.

Also note that BTC simply doens't include the fee and instead just computes it as the difference between outputs vs input and this saves some space. In this case we could just have Type on tx and no type for slips. I'm probably missing somethign around how fees are paid out, so perhaps that's what "Fee" is...? perhaps we could call it goldenFee or something if that's the case.

Reply:
This would be great for @trevelyan to weigh in on, but I agree it would be nice to simply this by just having either or.